File: app/src/main/java/com/cfks/goosedroid/GooseDesktop/HttpLlmBackend.java
package com.cfks.goosedroid.GooseDesktop;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * DEV ONLY — เชื่อมไปยัง llama-server บน PC (เช่น http://192.168.1.x:8080)
 * ใช้ endpoint /completion (raw prompt) เพื่อคุม chatml format เอง
 * ต้องถูกตัดออกก่อนปล่อย release (Phase 7 จะสลับเป็น PetLlama)
 */
public final class HttpLlmBackend implements LlmBackend {
    private final String baseUrl;
    private volatile HttpURLConnection lastConnection;

    public HttpLlmBackend(String baseUrl) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
    }

    @Override
    public String generate(String prompt, int maxTokens, float temp, float repeatPenalty) {
        try {
            JSONObject body = new JSONObject()
                    .put("prompt", prompt)
                    .put("n_predict", maxTokens)
                    .put("temperature", temp)
                    .put("repeat_penalty", repeatPenalty)
                    .put("stop", new JSONArray().put("<|im_end|>"));

            HttpURLConnection c = (HttpURLConnection) new URL(baseUrl + "/completion").openConnection();
            c.setRequestMethod("POST");
            c.setDoOutput(true);
            c.setRequestProperty("Content-Type", "application/json");
            c.setConnectTimeout(4000);
            c.setReadTimeout(60000);
            synchronized (this) { lastConnection = c; }

            try (OutputStream os = c.getOutputStream()) {
                os.write(body.toString().getBytes(StandardCharsets.UTF_8));
            }

            int code = c.getResponseCode();
            if (code != 200) { c.disconnect(); return null; }

            StringBuilder sb = new StringBuilder();
            try (BufferedReader r = new BufferedReader(
                    new InputStreamReader(c.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = r.readLine()) != null) sb.append(line);
            }
            c.disconnect();
            return new JSONObject(sb.toString()).optString("content", null);
        } catch (Exception e) {
            return null;
        }
    }

    @Override
    public void shutdown() {
        try { HttpURLConnection c = lastConnection; if (c != null) c.disconnect(); } catch (Exception ignored) {}
    }
}

File: app/src/main/java/com/cfks/goosedroid/brain/HttpLlmBackend.kt
package com.cfks.goosedroid.brain

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class HttpLlmBackend(
    var endpointUrl: String = "http://127.0.0.1:8080/completion"
) : LlmBackend {

    private val TAG = "HttpLlmBackend"

    override val backendName: String = "llama-server (Dev HTTP)"
    override val backendType: LlmBackendType = LlmBackendType.LOCAL_SERVER

    override fun isAvailable(): Boolean {
        return endpointUrl.isNotBlank() && (endpointUrl.startsWith("http://") || endpointUrl.startsWith("https://"))
    }

    override fun getStatusDetails(): String = "Endpoint: $endpointUrl"

    override suspend fun generate(
        prompt: String,
        maxTokens: Int,
        temp: Float,
        repeatPenalty: Float
    ): String? = withContext(Dispatchers.IO) {
        try {
            val url = URL(endpointUrl)
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json; utf-8")
            conn.setRequestProperty("Accept", "application/json")
            conn.connectTimeout = 3500
            conn.readTimeout = 8000
            conn.doOutput = true

            val payload = JSONObject().apply {
                put("prompt", prompt)
                put("n_predict", maxTokens)
                put("temperature", temp.toDouble())
                put("repeat_penalty", repeatPenalty.toDouble())
                put("stop", listOf("<|im_end|>", "<|endoftext|>"))
            }

            OutputStreamWriter(conn.outputStream).use { writer ->
                writer.write(payload.toString())
                writer.flush()
            }

            val responseCode = conn.responseCode
            if (responseCode == HttpURLConnection.HTTP_OK) {
                BufferedReader(InputStreamReader(conn.inputStream)).use { reader ->
                    val responseStr = reader.readText()
                    val respJson = JSONObject(responseStr)
                    return@withContext respJson.optString("content", respJson.optString("text", null))
                }
            } else {
                Log.w(TAG, "HTTP server returned code: $responseCode")
                null
            }
        } catch (e: Exception) {
            Log.d(TAG, "llama-server not reachable (${e.message})")
            null
        }
    }
}
